File: internal/tools/read_symbols.go
package tools

import (
	"context"
	"fmt"
	"strings"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
	"github.com/nanami7777777/gitlab-code-reader-mcp/internal/gitlab"
	"github.com/nanami7777777/gitlab-code-reader-mcp/internal/symbols"
)

func ReadSymbols(client *gitlab.Client) (mcp.Tool, server.ToolHandlerFunc) {
	tool := mcp.NewTool("gl_read_symbols",
		mcp.WithDescription("Extract code symbols (functions, classes, interfaces) from a file. For small files (<300 lines) returns full content. For large files returns signatures with line numbers. Like Claude Code's readCode tool."),
		mcp.WithString("project_id", mcp.Required(), mcp.Description("GitLab project ID or path")),
		mcp.WithString("file_path", mcp.Required(), mcp.Description("Path to the file")),
		mcp.WithString("ref", mcp.Description("Branch, tag, or commit SHA")),
		mcp.WithString("symbol_filter", mcp.Description("Filter symbols by name (case-insensitive substring match)")),
	)

	handler := func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		args := getArgs(req.Params.Arguments)
		projectID := argStr(args, "project_id")
		filePath := argStr(args, "file_path")

		ref := argStr(args, "ref")
		if ref == "" {
			r, err := client.DefaultBranch(projectID)
			if err != nil {
				return mcp.NewToolResultError(fmt.Sprintf("❌ Error: %v", err)), nil
			}
			ref = r
		}

		file, err := client.GetFileContent(projectID, filePath, ref)
		if err != nil {
			return mcp.NewToolResultError(guidedError(err, "read_symbols", args)), nil
		}

		if isBinary(file.Content) {
			return mcp.NewToolResultText(fmt.Sprintf("Binary file: %s (%s)", filePath, formatSize(file.Size))), nil
		}

		content, err := decodeBase64(file.Content)
		if err != nil {
			return mcp.NewToolResultError(fmt.Sprintf("❌ Decode error: %v", err)), nil
		}

		lines := strings.Split(content, "\n")
		totalLines := len(lines)
		header := fmt.Sprintf("File: %s (%s, %d lines, ref: %s)", filePath, formatSize(file.Size), totalLines, ref)

		// Small file: return full content
		if totalLines <= 300 {
			for i, l := range lines {
				lines[i] = truncateLine(l, 500)
			}
			numbered := addLineNumbers(strings.Join(lines, "\n"), 1)
			return mcp.NewToolResultText(fmt.Sprintf("%s\n(small file — returning full content)\n%s\n%s", header, strings.Repeat("─", 60), numbered)), nil
		}

		// Large file: extract symbols
		syms := symbols.Extract(content, filePath)

		filter := argStr(args, "symbol_filter")
		if filter != "" {
			filterLower := strings.ToLower(filter)
			var filtered []symbols.Symbol
			for _, s := range syms {
				if strings.Contains(strings.ToLower(s.Name), filterLower) {
					filtered = append(filtered, s)
				}
			}
			syms = filtered
		}

		output := symbols.Format(syms)
		return mcp.NewToolResultText(fmt.Sprintf("%s\n(large file — showing %d symbol signatures)\n%s\n\n💡 Use gl_read_file with start_line/end_line to read specific implementations.", header, len(syms), output)), nil
	}

	return tool, handler
}

File: src/tools/read-symbols.ts
import { z } from "zod";
import type { GitLabClient } from "../gitlab/client.js";
import { isBinaryContent, decodeFileContent, formatSize } from "../utils/format.js";
import { extractSymbols, formatSymbols } from "../utils/symbols.js";
import { addLineNumbers, truncateLine } from "../utils/format.js";

export const readSymbolsSchema = z.object({
  project_id: z.string().describe("GitLab project ID or path"),
  file_path: z.string().describe("Path to the file"),
  ref: z.string().optional().describe("Branch, tag, or commit SHA"),
  symbol_filter: z.string().optional().describe("Filter symbols by name (case-insensitive substring match)"),
});

export type ReadSymbolsInput = z.infer<typeof readSymbolsSchema>;

/**
 * Inspired by Claude Code's readCode tool:
 * - Small files (<300 lines): return full content
 * - Large files: return symbol signatures with line numbers
 */
export async function readSymbols(client: GitLabClient, input: ReadSymbolsInput): Promise<string> {
  const ref = input.ref ?? (await client.getDefaultBranch(input.project_id));
  const file = await client.getFileContent(input.project_id, input.file_path, ref);

  if (isBinaryContent(file.content)) {
    return `Binary file: ${input.file_path} (${formatSize(file.size)})`;
  }

  const content = decodeFileContent(file.content);
  const lines = content.split("\n");
  const totalLines = lines.length;

  const header = `File: ${input.file_path} (${formatSize(file.size)}, ${totalLines} lines, ref: ${ref})`;

  // Small file: return full content (like Claude Code's readCode for <10k chars)
  if (totalLines <= 300) {
    const processed = lines.map((l) => truncateLine(l, 500));
    const numbered = addLineNumbers(processed.join("\n"), 1);
    return `${header}\n(small file — returning full content)\n${"─".repeat(60)}\n${numbered}`;
  }

  // Large file: extract and return symbols
  let symbols = extractSymbols(content, input.file_path);

  if (input.symbol_filter) {
    const filter = input.symbol_filter.toLowerCase();
    symbols = symbols.filter((s) => s.name.toLowerCase().includes(filter));
  }

  const symbolOutput = formatSymbols(symbols);

  return `${header}\n(large file — showing ${symbols.length} symbol signatures)\n${symbolOutput}\n\n💡 Use gl_read_file with start_line/end_line to read specific implementations.`;
}
